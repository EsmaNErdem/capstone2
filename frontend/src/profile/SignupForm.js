import React, { useState } from 'react';
import { Link, useNavigate } from 'react-router-dom';
import { Formik, Field, Form, ErrorMessage } from 'formik';
import { Typography } from '@mui/material';
import Alert from "../utilities/Alert";
import Loading from "../utilities/Loading";
import * as Yup from 'yup';
import "./FormContainer.css"

/**
 * SignUpForm Component
 * Displays form and renders controlled components functionality
 * Provides form validation with formik
 * 
 * A form component for user registration. Handles user input for username, password, first name, last name, email and user image.
 * Displays form errors if registration fails.
 * On submission:
 * - calls signup function prop
 * - redirects to / route
 *
 * Routes ==> SignupForm ==> Alert
 */
const SignupForm = ({ signup }) => {
    console.debug("SignupForm");

    const navigate = useNavigate();
    const [error, setError] = useState(null);
    const [loading, setLoading] = useState(false)

    /**
     * Handles form submission. Calls the signup function with the form data.
     * Redirects to the home page upon successful registration, or displays form errors.
     */
    const handleSubmit = async (values, { setSubmitting }) => {
        try {
            setLoading(true);
            const result = await signup(values);
    
            if (result.success) {
                navigate("/");
                setLoading(false)
            } else {
                setError(result.error);
            }
        } catch (error) {
            setError("Please use another username. This username is taken. ");
        } finally {
            setSubmitting(false);
        }
    };

    return (
        <div>
            <Typography variant="h2" sx={{  marginBottom: 2 }}>Sign Up</Typography>
            {error ? <Alert type="danger" messages={[error]} />: null}
            <Formik
                initialValues={{ username: '', password: '', email: '', img: '' }}
                validationSchema={Yup.object({
                    username: Yup.string()
                    .max(35, 'Must be 35 characters or less')
                    .required('Username is required'),
                    password: Yup.string()
                    .max(20, 'Must be 20 characters or less')
                    .required('Password is required'),
                    email: Yup.string()
                    .email('Invalid email address')
                    .required('Email is required'),
                    img: Yup.string().notRequired(),
                })}
                onSubmit={handleSubmit}
                >
                <Form className='FormContainer'>
                    <div>
                        <label htmlFor="username">Username</label>
                        <Field id="username" name="username" type="text" placeholder="Username" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="username"/></span>
                    </div>

                    <div>
                        <label htmlFor="password">Password</label>
                        <Field id="password" name="password" type="password" placeholder="Password" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="password" /></span>
                    </div>

                    <div>
                        <label htmlFor="email">Email Address</label>
                        <Field id="email" name="email" type="email" placeholder="Email" />
                        <span className='ErrorMessageContainer'><ErrorMessage name="email" /></span>
                    </div>

                    <div>
                        <label htmlFor="img">Profile Image URL (optional)</label>
                        <Field id="img" name="img" type="text" placeholder="Profile Image URL (optional)"/>
                        <span className='ErrorMessageContainer'><ErrorMessage name="img" /></span>
                    </div>

                    {/* <div>
                        <label htmlFor="userImg">Profile Image</label>
                        <Field name="userImg" type="file" encType="multipart/form-data"/>
                        <input type="file" name="userImg" encType="multipart/form-data"/>
                        <span className='ErrorMessageContainer'><ErrorMessage name="profileImage" /></span>
                    </div> */}

                    <button type="submit">Submit</button>
                    {loading && <Loading style={{ color:"orangered" }} />}                
                </Form>
            </Formik>
            <Link to={`/login`} data-testid="user-login-link">
                <p>Login instead?</p>
            </Link>
        </div>
    );
};

export default SignupForm;
