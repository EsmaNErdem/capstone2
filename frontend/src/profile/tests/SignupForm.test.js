import { render, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom';
import SignupForm from "../SignupForm";

test("renders without crashing", async () => {
    render(
        <MemoryRouter >
            <SignupForm />
        </MemoryRouter>
    ); 
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(
        <MemoryRouter >
            <SignupForm />
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

test("SignUpForm renders correctly", () => {
    const { getByLabelText, getByText, getByTestId } = render(
        <MemoryRouter >
            <SignupForm />
        </MemoryRouter>
    );
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const emailInput = getByLabelText("Email Address");
    const submitButton = getByText("Submit");
    
    const profileLink = getByTestId('user-login-link');
    expect(profileLink.getAttribute('href')).toBe('/login');

    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });
  
  test("Submitting the form calls signup function", async () => {
    const mockSignUp = jest.fn().mockResolvedValue({ success: true }); 

    const { getByLabelText, getByText } = render(
          <MemoryRouter >
              <SignupForm signup={mockSignUp} />
          </MemoryRouter>
      );
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const emailInput = getByLabelText("Email Address");
    const submitButton = getByText("Submit");
  
    fireEvent.change(usernameInput, { target: { value: "username" } });
    fireEvent.change(passwordInput, { target: { value: 123456789 } });
    fireEvent.change(emailInput, { target: { value: "user@email.com" } });
    fireEvent.click(submitButton);
  
    await waitFor(() => {
        expect(mockSignUp).toHaveBeenCalledWith({
          username: "username",
          password: "123456789",
          email: "user@email.com",
          img: ""
        });
    });
  });
   
