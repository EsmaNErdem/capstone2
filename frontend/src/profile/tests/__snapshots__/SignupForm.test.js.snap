// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="MuiTypography-root MuiTypography-h2 css-vqgaav-MuiTypography-root"
    >
      Sign Up
    </h2>
    <form
      action="#"
      class="FormContainer"
    >
      <div>
        <label
          for="username"
        >
          Username
        </label>
        <input
          id="username"
          name="username"
          placeholder="Username"
          type="text"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="email"
        >
          Email Address
        </label>
        <input
          id="email"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <div>
        <label
          for="img"
        >
          Profile Image URL (optional)
        </label>
        <input
          id="img"
          name="img"
          placeholder="Profile Image URL (optional)"
          type="text"
          value=""
        />
        <span
          class="ErrorMessageContainer"
        />
      </div>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
    <a
      data-testid="user-login-link"
      href="/login"
    >
      <p>
        Login instead?
      </p>
    </a>
  </div>
</DocumentFragment>
`;
